example_allocators/compartment_alloc/compartment_alloc.c

```diff
@@ -0,0 +1,91 @@
+#include <cheriintrin.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include "compartment_alloc.h"
+#include "utils.h"
+
+heap_compartment compartments[maxCompartments];
+int numCompartments = 0;
+
+/*
+ * Initialise a compartment with `size_in_bytes` space, where all future
+ * allocations will be constrained by `dc` as if it was the DDC when allocation
+ * occurred.
+ */
+void
+*__capability init_compartment(int size_in_bytes, void *__capability dc)
+{
+    // We impose a limit on our # of compartments
+    if (numCompartments + 1 == maxCompartments) {
+        return NULL;
+    }
+
+    // Allocate memory for this compartment
+    void *buf = mmap(NULL, size_in_bytes, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1, 0);
+    if (buf == MAP_FAILED)
+    {
+        perror("error in initial mem allocation");
+        exit(-1);
+    }
+
+    // Create the new compartment
+    compartments[numCompartments] = (heap_compartment){
+        buf,                     // Space on the heap
+        0,                       // bytes allocated (currently none!)
+        size_in_bytes,           // Our maximum size, to check when allocating
+        dc,                      // The data capability to allocate against (like a DDC)
+        sealed_reference(buf)    // An identifier which allows a holder to allocate in this component
+    };
+
+    // Return the sealed capability which acts as a token of authentication for a component.
+    return compartments[numCompartments++].identifier;
+}
+
+/*
+ * Allocate `len` bytes in the compartment identified by `compartment_id`.
+ * if `compartment_id` doesn't exist, returns NULL
+ */
+void *__capability malloc_compartment(int len, void *__capability compartment_id)
+{
+    size_t rounded_length;
+    size_t new_allocated;
+    heap_compartment compartment = compartments[0];
+    void *__capability allocated;
+    void *addr;
+    int i;
+
+    // Search for a compartment with the given identifier
+    for (i = 0; i < maxCompartments && compartments[i].identifier != compartment_id; i++){}
```

Comment:
Not quite sure what's happening here. Is there meant to be no semicolon after the empty body of this for? Can't we omit the `{}` completely and just leave the for itself?

Also, did you check that edge case if the sought-for compartment is at the end of `compartments`?

Reply:
An empty body and a semicolon immediately following the `for` brackets will do the same thing — I'm not sure what the idiomatic C would be here, so happy to follow whatever our convention is for something like this. Would you prefer to see `for (i = 0; i < maxCompartments && compartments[i].identifier != compartment_id; i++);`, or would you rewrite it another way?

Reply:
It just looks confusing to me. If you think it's fine, leave as is. I just prefer being more verbose.

Reply:
There's no great answer here -- personally, I think the `{}` is the least worst option (it's what I do in this situation).